echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/notification/InteractiveNotificationCallbackHandler.groovy

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2020 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.notification
+
+import com.jakewharton.retrofit.Ok3Client
+import com.netflix.spinnaker.echo.api.Notification
+import com.netflix.spinnaker.echo.api.Notification.InteractiveActionCallback
+import com.netflix.spinnaker.retrofit.Slf4jRetrofitLogger
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.core.env.Environment
+import org.springframework.http.HttpHeaders
+import org.springframework.stereotype.Component
+import retrofit.RestAdapter
+import retrofit.client.Response
+import retrofit.converter.JacksonConverter
+import retrofit.http.Body
+import retrofit.http.Header
+import retrofit.http.POST
+
+import static retrofit.Endpoints.newFixedEndpoint
+
+@Slf4j
+@Component
+class InteractiveNotificationCallbackHandler {
+  @Autowired
+  Ok3Client spinnakerServiceClient
+
+  @Autowired
+  List<InteractiveNotificationService> notificationServices
+
+  @Autowired
+  private Environment environment
+
+  /**
+   * Processes a callback request from Slack by relaying it to the downstream service which originated the message
+   * referenced in the payload.
+   * @param event
+   */
+  Map processCallback(String source, Map content, HttpHeaders headers) {
+    log.debug("Received interactive notification callback request from ${source}")
+
+    InteractiveNotificationService notificationService = notificationServices.stream()
+      .find { it.supportsType(Notification.Type.valueOf(source.toUpperCase())) }
+
+    if (notificationService == null) {
+      throw new IllegalArgumentException("NotificationService for ${source} no registered")
+    }
+
+    InteractiveActionCallback callback = notificationService.parseInteractionCallback(content)
+    String requestUrl = environment.getProperty("${callback.serviceId}.baseUrl") + "/notifications/callback"
+    SpinnakerService spinnakerService = new RestAdapter.Builder()
+      .setEndpoint(newFixedEndpoint(requestUrl))
+      .setClient(spinnakerServiceClient)
+      .setLogLevel(RestAdapter.LogLevel.BASIC)
+      .setLog(new Slf4jRetrofitLogger(SpinnakerService.class))
+      .setConverter(new JacksonConverter())
+      .build()
+      .create(SpinnakerService.class)
```

Comment:
This allows us to dynamically get a client for any of the Spinnaker services in the config, without having to hard-code names and have a `switch` statement.

Reply:
These service objects are going to be something you'll want to cache. Creating new retrofit clients is not a cheap operation.

Reply:
I had a hunch about that, but was waiting for the pros to confirm. 😉 

Would instantiating just the `Retrofit.Builder` in Spring config and then building the service object here be a good-enough compromise, or is the introspection of the service class in `create(SpinnakerService.clas)` that is costly?